File: javascript/array/121_Best_Time_to_Buy _and_Sell_Stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let l = 0;
    let r = 1;
    let maxProfit = 0;

    while (r < prices.length) {
        if (prices[l] < prices[r]) {
            profit = prices[r] - prices[l];
            maxProfit = Math.max(profit, maxProfit);
        }
        else l = r;
        r ++;
    }
    return maxProfit;
};

output1 = maxProfit(prices = [7,1,5,3,6,4])  // 5
output2 = maxProfit(prices = [7,6,4,3,1])  // 0
console.log(output1, output2)

File: python/two_pointers/121_Best_Time_to_Buy _and_Sell_Stock.py
'''
Time: O(n)
Space: O(1)
'''
from typing import List

class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        l, r = 0, 1  # l (left) = buy; r (right) = sell
        max_profit = 0
        while r < len(prices):
            if prices[l] < prices[r]:
                profit = prices[r] - prices[l]
                max_profit = max(max_profit, profit)
            else:
                l = r
            r += 1

        return max_profit


output1 = Solution().maxProfit(prices = [7,1,5,3,6,4])  # 5
output2 = Solution().maxProfit(prices = [7,6,4,3,1])  # 0
print(output1, output2)